### PR TITLE
feature(security) authentification endpoints

### PR DESCRIPTION
**Pull request by AlexTorianyk**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/igroup-opole/yesterday-api/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

## Other information

### CODE REVIEW

File: YesterdayApi/Core/Email/Builder/Builders/EmailBuilderBase.cs [hunk 1]
```diff
@@ -15,7 +16,7 @@ public abstract class EmailBuilderBase
 
         public abstract void ResetEmail();
         public abstract void AppendUserDetails(IEmailReceiver details);
-        public abstract void PrepareBody();
+        public abstract Task PrepareBody();
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Email/Builder/Builders/EmailBuilderBase.cs]: Good catch!

File: YesterdayApi/Core/Identity/IdentityService.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+using Microsoft.Extensions.Configuration;
+using YesterdayApi.Core.Identity.Web;
+using YesterdayApi.Core.Users;
+using YesterdayApi.Utilities.Exceptions;
+
+namespace YesterdayApi.Core.Identity
+{
+    public class IdentityService : IIdentityService
+    {
+        private const string DiscoveryUrl = "http://localhost:5000";
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/IdentityService.cs]: Maybe would be better to put this url into the configuration like connection string?

File: YesterdayApi/Core/Identity/IdentityService.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        public async Task<IdentityResult> RegisterUser(UserRegistrationRequest userRegistrationRequest)
+        {
+            var newUser = _mapper.Map<User>(userRegistrationRequest);
+            var result = await _userIdentityService.CreateAsync(newUser, userRegistrationRequest.Password);
+
+            return result;
+        }
+
+        public async Task SetAdmin(string userName)
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/IdentityService.cs]: Why do you prefer using username instead of id? Id is already indexed by db so it's much faster.

File: YesterdayApi/Core/Identity/IdentityService.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                throw new BadRequestException("Error", "User is already in role.");
+            }
+
+            await _userIdentityService.AddToRoleAsync(user, "Admin").ConfigureAwait(false);
+        }
+
+        private async Task<DiscoveryResponse> GetDiscoveryDocument()
+        {
+            var disco = await _client.GetDiscoveryDocumentAsync(DiscoveryUrl);
+            if (disco.IsError)
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/IdentityService.cs]: Be consistent - in previous method you have empty line before statement 🚀

File: YesterdayApi/Core/Identity/TokenDecoder.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        public int GetUserId(string authHeader)
+        {
+            if (string.IsNullOrEmpty(authHeader))
+            {
+                throw new UnauthorizedException("Error", "Authorization header is empty.");
+            }
+
+            var decodedToken = GetDecodedToken(authHeader);
+
+            if (TokenTooOld(decodedToken.ValidTo))
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/TokenDecoder.cs]: Maybe method IsTokenTooOld or IsTokenExpired?

File: YesterdayApi/Core/Identity/TokenDecoder.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        private static JwtSecurityToken GetDecodedToken(string authHeader)
+        {
+            var encodedJwt = authHeader.Substring(7);
+            var decodedToken = new JwtSecurityToken(encodedJwt);
+            return decodedToken;
+        }
+
+        private static bool TokenTooOld(DateTime expiryDate)
+        {
+            return (DateTime.Now.Date - expiryDate).TotalDays > 5;
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/TokenDecoder.cs]: Why 5? What's the reason?

File: YesterdayApi/Core/Identity/Validators/UserIdentityRequestValidator.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+using YesterdayApi.Utilities.AutomaticDI;
+
+namespace YesterdayApi.Core.Identity.Validators
+{
+    public class UserIdentityRequestValidator : AbstractValidator<UserIdentityRequest>, ITransient
+    {
+        public UserIdentityRequestValidator()
+        {
+            RuleFor(ur => ur.Password).NotNull();
+            RuleFor(ur => ur.Password).Length(5, 50);
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/Validators/UserIdentityRequestValidator.cs]: Correct me if I'm wrong but I think you can put two requirements in one line like:

```c#
RuleFor(ur => ur.Password).NotNull().Length(5, 50);
```

File: YesterdayApi/Core/Identity/Validators/UserRegistrationRequestValidator.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace YesterdayApi.Core.Identity.Validators
+{
+    public class UserRegistrationRequestValidator : AbstractValidator<UserRegistrationRequest>, ITransient
+    {
+        public UserRegistrationRequestValidator()
+        {
+            Include(new UserIdentityRequestValidator());
+            RuleFor(urr => urr.Email).NotNull();
+            RuleFor(urr => urr.Email).Length(7, 230);
```
Comment (rafalschmidt97) [inline YesterdayApi/Core/Identity/Validators/UserRegistrationRequestValidator.cs]: Weird validator. Wouldn't be this better if you'll try this? What's about "x@a.pl (there's a domain like that 🤣)

```c#
RuleFor(urr => urr.Email).NotEmpty().MaxLength(230);
```

File: YesterdayApi/Startup.cs [hunk 9]
```diff
@@ -25,9 +26,19 @@ public Startup(IConfiguration configuration)
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
+
+            services.AddAuthentication("Bearer")
+                .AddIdentityServerAuthentication(options =>
+                {
+                    options.Authority = "http://localhost:5000";
```
Comment (rafalschmidt97) [inline YesterdayApi/Startup.cs]: Maybe would be better to put this url into the configuration?

File: YesterdayApi/Core/Identity/IdentityService.cs (hunk 2 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/IdentityService.cs]: Idk, i think it's specific to this one class, so it kinda makes sense to keep it cohesive.

File: YesterdayApi/Core/Identity/IdentityService.cs (hunk 3 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/IdentityService.cs]: How would an admin know someone else's id? I feel like username is more accessible to the admin, so I chose to use it, but if you provide a method of finding out a user's id, I'll gladly change it 👍

File: YesterdayApi/Core/Identity/TokenDecoder.cs (hunk 5 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/TokenDecoder.cs]: I check if the token expired less than 5 days ago, because i want to keep the access token lifetime quite short, so it's different things.

File: YesterdayApi/Core/Identity/TokenDecoder.cs (hunk 6 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/TokenDecoder.cs]: I just chose a random number and went with it.

File: YesterdayApi/Core/Identity/Validators/UserIdentityRequestValidator.cs (hunk 7 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/Validators/UserIdentityRequestValidator.cs]: Idk i think it's better to separate separate rules in different lines so they don't get mixed up.

File: YesterdayApi/Core/Identity/IdentityService.cs (hunk 2 above)
Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/IdentityService.cs]: Yeah nvm, found your second comment on this string down the line, nice catch.

File: YesterdayApi/Core/Identity/IdentityService.cs (hunk 3 above)
Comment (rafalschmidt97) [reply YesterdayApi/Core/Identity/IdentityService.cs]: Admin almost always need to get users (like list them) before setting them admin privileges - through some ui. Otherwise we'll still do this from sql query. But whatever 😇

File: YesterdayApi/Startup.cs (hunk 9 above)
Comment (AlexTorianyk) [reply YesterdayApi/Startup.cs]: Yeah.

Comment (AlexTorianyk) [reply YesterdayApi/Startup.cs]: Yeah.

File: YesterdayApi/Core/Identity/TokenDecoder.cs (hunk 5 above)
Comment (rafalschmidt97) [reply YesterdayApi/Core/Identity/TokenDecoder.cs]: Can we instead of this shorten lifetime of token (i mean expired date)?

Comment (AlexTorianyk) [reply YesterdayApi/Core/Identity/TokenDecoder.cs]: Can you rephrase please, I don't get what you mean.

Comment (rafalschmidt97) [reply YesterdayApi/Core/Identity/TokenDecoder.cs]: Ahhh... I've got idea right now haha. Nvm. It's fine
